src/System.Collections/tests/Generic/Dictionary/Dictionary.Generic.Tests.ConcurrentAccessDetection.netcoreapp.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Generic.Dictionary
+{
+    public class DictionaryConcurrentAccessDetectionTests
+    {
+        async static Task DictionaryConcurrentAccessDetection<TKey, TValue>(Dictionary<TKey, TValue> dictionary, bool isValueType, object comparer, Action<Dictionary<TKey, TValue>> add, Action<Dictionary<TKey, TValue>> get, Action<Dictionary<TKey, TValue>> remove, Action<Dictionary<TKey, TValue>> removeOutParam)
+        {
+            Task task = Task.Factory.StartNew(() =>
+            {
+                //break internal state
+                var entriesType = dictionary.GetType().GetField("_entries", BindingFlags.NonPublic | BindingFlags.Instance);
+                var entriesInstance = (Array)entriesType.GetValue(dictionary);
+                var field = entriesInstance.GetType().GetElementType();
+                var entryArray = (Array)Activator.CreateInstance(entriesInstance.GetType(), new object[] { ((IDictionary)dictionary).Count });
+                var entry = Activator.CreateInstance(field);
+                entriesType.SetValue(dictionary, entryArray);
+
+                Assert.Equal(comparer, dictionary.GetType().GetField("_comparer", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(dictionary));
+                Assert.Equal(isValueType, dictionary.GetType().GetGenericArguments()[0].IsValueType);
+                Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => add(dictionary)).TargetSite.Name);
+                Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => get(dictionary)).TargetSite.Name);
+                //Remove is not resilient yet
+                //Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => remove(dictionary)).TargetSite.Name);
+                //Assert.Equal("ThrowInvalidOperationException_ConcurrentOperationsNotSupported", Assert.Throws<InvalidOperationException>(() => removeOutParam(dictionary)).TargetSite.Name);
+            }, TaskCreationOptions.LongRunning);
+
+            //Wait max 60 seconds, could loop forever
+            Assert.True((await Task.WhenAny(task, Task.Delay(TimeSpan.FromSeconds(60))) == task) && task.IsCompletedSuccessfully);
+        }
+
+        [Theory]
+        [InlineData(null)]
+        [InlineData(typeof(CustomEqualityComparerInt32ValueType))]
+        async public static Task Add_DictionaryConcurrentAccessDetection_ValueTypeKey(Type comparerType)
```

Comment:
ok i used static because i'm working also on Span tests refactor...and in that project are all static..is there a guideline?

https://github.com/dotnet/corefx/blob/master/src/System.Memory/tests/ReadOnlySpan/SequenceEqual.char.cs

Reply:
I and most of us prefer non-static test methods.

Reply:
@karelz should we have a guideline for static vs non-static test methods?

Reply:
>I and most of us prefer non-static test methods.

Also me and my team.

Reply:
I think Xunit runs the ctor for each test. So non static gives you a fresh instance. Otherwise I'm not sure we care, save consistency. @stephentoub?

Reply:
Correct: https://xunit.github.io/docs/shared-context#constructor

Reply:
So static class for tests are more performant.